### ISSUE TITLE
New client-server protocol for better light client support

### ISSUE THREAD

**Issue by ebuchman** [tendermint/spec#202]
Model based testing recently surfaced an [issue](https://github.com/informalsystems/tendermint-rs/issues/650) that, while possibly not a bug per se (depending on how you look at it), suggests we could improve the client-server communication in the light client protocol.

The issue in the test was from a signed header with an empty commit AND without enough trust. We marked it as "not enough trust" (ie. start bisection) but the test expected it to fail as "invalid commit". It's trivial to just check for empty commits first, but the issue is deeper than that - the full node could provide a bogus validator set and commit that verifies correctly but is not actually from the blockchain, and we won't know until we go through bisection. 

This surfaces a confusion over whether the commit should be verified before or after checking if the header can be trusted to skip to in the first place. This confusion is unfortunate. 

An alternative would be to expose a new endpoint on full nodes where light clients can submit their last trusted header and the full node can do the bisection for them! Full nodes won't have to check signatures (since they're already verified), they'd just be looking at validator set overlap. Then they'd return a signed header that the light client should definitely be able to trust. 

This is in contrast to the light client scheduling the bisection itself and being at the mercy of full nodes giving it wonky headers. With the new approach, clients would be able to detect faulty full nodes immediately, since the only acceptable response is a signed header they can trust.

This would also make the light node function more like the on chain IBC handler, which can't schedule its own bisection and is at the mercy of doing one-step validation of whatever headers it's sent. 

The downside of this proposal is of course an extra endpoint and a bit more complexity on the full node, but seems like it may be worth it.

**Comment by cmwaters** [comment]
Yeah I thought about this before. If the full node knows the trusted and target headers it can work out the minimum trace of light blocks it needs to send to the light client for it to validate the target header and could either send them one at a time per request or all at once. That way we potentially save on a bunch of network calls for intermediate headers that might not be helpful. 

> With the new approach, clients would be able to detect faulty full nodes immediately, since the only acceptable response is a signed header they can trust.

I don't quite understand this. The light client would still need to check both that the validator set met the required trust level and that 2/3 of the validator set signed the header.

**Comment by konnov** [comment]
> I don't quite understand this. The light client would still need to check both that the validator set met the required trust level and that 2/3 of the validator set signed the header.

I also think that for the resulting sequence of fetched blocks the light client will still have to do the 2/3s check against the commits and the 1/3 check. A malicious full node can also produce a complete sequence of blocks from the trusted height to the target height and make it different in the very last block, so the light client has to do complete sequential check and fail in the end.

**Comment by andrey-kuprianov** [comment]
I will try to comment up to my limited understanding; Ethan will correct me if I am wrong. 

What we construct here, is the meta-algorithm: whatever checks the current light client does, we just take all the checks on the happy path branch, those that allow to trust the new block, and _require_ from the full node to provide us the block with the maximal height that satisfies those conditions. _Whatever_ checks they are, the result should be verifiable in 1 step; otherwise we will know the full node is malicious.

The only damage the full node can cause under such a meta-algorithm, is to try slow the light client down, i.e. to provide a block with a smaller height than possible. But even this misbehavior can be detected via cross-checking with other full nodes. In particular, it should not be possible to fool the light client and make it trust any sequence of fake blocks, all validation happens in 1 step.

**Comment by ebuchman** [comment]
Exactly. We definitely still have the do the +1/3 trust check and the +2/3 commit validation - by one step I just meant that we wouldn't need any intermediate headers to verify a header since the full node would always be expected to give us headers we could immediately verify - even if it gave us a list of them we could verify in order. The idea is to avoid the case where we're given a header that fails the +1/3 trust check. 

And yes we can always cross reference so if one node is trying to force us to go sequential we can see if other nodes are offering a skip and if so immediately validate if their skip is legit or not by doing the +1/3 and +2/3 checks for the one header without needing to do bisection ourselves :)

**Comment by josef-widder** [comment]
I suspect that a lot of this discussion is about code that is rarely executed under the current Cosmos Hub: no validator set change = no bisection. As a result, my feeling in this discussion is that we try to optimize something, but I don't clearly see what is the space we are operating in and what are we trying to optimize. Here is am menu of scenarios/goals we might optimize for:

- the lightclient is rarely is called (it is more likely to need bisection)
- the lightclient is called every minute (it is always roughly synchronized to the blockchain. Might never need bisection)
- rare lightclient attack (expensive extra computation is acceptable)
- regular lightclient attack (we need to optimize for it)
- frequent validator set changes
- rare validator set changes
- be efficient in the lucky path
- minimize extra work because of buggy/stupid full nodes (empty commit)
- minimize extra work due to clever adversarial full node (with or without lightclient attack; btw. what is the incentive of a full node to just slow down the lightclient by extra work?)
- talk to possibly few full nodes (only one)
- cross check everything with multiple full nodes
- assume we are always connected to at least one correct full node
- assume we are from time to time connected to at least one correct full node

I don't think this list is by any means exhaustive. I suspect different combinations of these goals may lead to different lightclient algorithms (that is, combination of some sort of "verification" and "attack detection"). I would like to understand what we try to achieve, then I can participate more constructively in the discussion ;-) 

Perhaps it might also make sense to use different lightclient solutions for different use cases. I guess a light client on a relayer that is online 24/7 and regularly updates its state should be optimized for a different scenario than a lightclient on a smartphone that is called once a day to check a specific transaction in a block.
